lib/board.rb

```diff
@@ -13,37 +13,35 @@ def add_mark(position, mark)
   end
 
   def available_positions
-    positions = []
-    grid.select { |cell| (0..8).to_a.include? cell}
+    grid.select { |cell| cell != :X && cell != :O}
   end
 
   def game_over?
-    winner || full? ? true : false
+    has_winner? || full?
   end
 
   def winner
     winner = false
     lines.each do |line|
-      winner = :X if all_x(line)
-      winner = :O if all_o(line)
+      return line[0] if all_same_marks?(line)
     end
     winner
   end
 
-  def empty?
-    !full?
+  def has_winner?
+    lines.any?{|line| all_same_marks?(line)}
   end
 
-  def full?
-    (grid & (0..8).to_a).empty?
+  def all_same_marks?(line)
+    line[0] == line[1] && line[1] == line[2]
   end
 
-  def all_x(line)
-    line.all? {|x| x == :X}
+  def empty?
+    !full?
   end
 
-  def all_o(line)
-    line.all? {|o| o == :O}
+  def full?
+    (grid & (0..8).to_a).empty?
```

Comment:
I'm using an intersection - the `&` compares two arrays and returns an array with the elements that both have in common. And in this case it should be empty.